## Wrangle and Analyze Data

Real-world data rarely comes clean. Using Python and its libraries, we will gather data from a variety of sources and in a variety of formats, assess its quality and tidiness, then clean it.


## Table of Contents
- [Introduction](#intro)
- [Part I - Gathering Data](#gather)
- [Part II - Assessing Data](#assess)
- [Part III - Cleaning Data](#clean)


<a id='intro'></a>
### Introduction

For this project I will be wrangling WeRateDogs Twitter data to create interesting and trustworthy analyses and visualizations. The Twitter archive contains only very basic tweet information. Additional gathering, then assessing and cleaning is required for "Wow!"-worthy analyses and visualizations. 

WeRateDogs is a Twitter account that rates people's dogs with a humorous comment about the dog. These ratings almost always have a denominator of 10. The numerators, though? Almost always greater than 10. 11/10, 12/10, 13/10, etc. Why? Because "they're good dogs Brent." WeRateDogs has over 4 million followers and has received international media coverage.

<a id='gather'></a>
#### Part I - Gathering Data

In this section I will gather data using 3 methodologies:
1. *The WeRateDogs Twitter archive*. __Downloaded manually__ from the server. 

2. *The tweet image predictions*, i.e., what breed of dog (or other object, animal, etc.) is present in each tweet according to a neural network. This file (image_predictions.tsv) is hosted on Udacity's servers and should be __downloaded programmatically__ using the Requests library and the following URL: https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv

3. Each tweet's retweet count and favorite ("like") count at minimum, and any additional data you find interesting. Using the tweet IDs in the WeRateDogs Twitter archive, __query the Twitter API__ for each tweet's JSON data __using Python's Tweepy library__ and store each tweet's entire set of JSON data in a file called *tweet_json.txt* file.


Let's start with importing neccessary libraries.

In [1]:
# Import libraries
import numpy as np
import pandas as pd
import requests
import os
import tweepy
import json

In [2]:
# Read WeRateDogs Twitter archive csv file
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
# First check if the data is properly loaded
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

#### Download tweet image predictions tsv file from the Udacity's server

url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url)

with open(url.split('/')[-1], 'wb') as f:
    f.write(r.content)

In [4]:
# Read the tweet image predictions tsv file
twitter_images = pd.read_csv('image-predictions.tsv', sep = '\t')

In [5]:
# First check if the data is properly loaded
twitter_images.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

#### Twitter authentification setup

consumer_key = ''
consumer_secret = ''
access_token = ''
access_secret = ''

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)

#### Download Tweets part 1

--- List for storing json outputs
tweet_list = []
--- List for storing Twitter IDs where the status has not been found
tweet_errors = []

--- Loop through Twitter IDs from the twitter archive file
--- Append json outputs to tweet_list and not found Twitter IDs to 'tweet_errors' list
for tweet_id in twitter_archive.tweet_id:
    try:
        tweet = api.get_status(tweet_id, tweet_mode='extended')
        tweet_list.append(tweet._json)
    except Exception as e:
        print(str(tweet_id)+ " _ " + str(e))
        tweet_errors.append(tweet_id)

#### Download Tweets part 2

--- Additional list for storing Twitter IDs where the status has not been found the second time
missing_list = []

--- Loop through the list of missing IDs
--- Append json outputs to tweet_list and not found Twitter IDs to missing_list
for missing_id in tweet_errors:
    try:
        tweet_list.append(tweet._json)
    except Exception as ex:
        print(str(missing_id)+ " _ " + str(ex))
        missing_list.append(ex)

#### Create and safe dataframe
--- Create DataFrames from list of dictionaries
json_df = pd.DataFrame(tweet_list)

--- Save the dataFrame in file
json_df.to_csv('tweet_json.txt', index=False)

In [6]:
# Read tweet_json.txt file
twitter_json = pd.read_csv('tweet_json.txt', encoding = 'utf-8')

In [7]:
# First check if the data is properly loaded
twitter_json.head()

contributors  coordinates                      created_at  \
0           NaN          NaN  Tue Aug 01 16:23:56 +0000 2017   
1           NaN          NaN  Tue Aug 01 00:17:27 +0000 2017   
2           NaN          NaN  Mon Jul 31 00:18:03 +0000 2017   
3           NaN          NaN  Sun Jul 30 15:58:51 +0000 2017   
4           NaN          NaN  Sat Jul 29 16:00:24 +0000 2017   

  display_text_range                                           entities  \
0            [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1           [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2           [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3            [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4           [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 892420639486877696, 'id_str'...           38126   
1  {'media': [{'id': 892177413194625024, 'id_str'...           32703   
2  {'media': [{'id': 891815175371796480, 'id_str'...           24623   
3  {'media': [{'id': 891689552724799489, 'id_str'...           41473   
4  {'media': [{'id': 891327551943041024, 'id_str'...           39639   

   favorited                                          full_text  geo  \
0      False  This is Phineas. He's a mystical boy. Only eve...  NaN   
1      False  This is Tilly. She's just checking pup on you....  NaN   
2      False  This is Archie. He is a rare Norwegian Pouncin...  NaN   
3      False  This is Darla. She commenced a snooze mid meal...  NaN   
4      False  This is Franklin. He would like you to stop ca...  NaN   

                         ...                          quoted_status  \
0                        ...                                    NaN   
1                        ...                                    NaN   
2                        ...                                    NaN   
3                        ...                                    NaN   
4                        ...                                    NaN   

   quoted_status_id quoted_status_id_str  quoted_status_permalink  \
0               NaN                  NaN                      NaN   
1               NaN                  NaN                      NaN   
2               NaN                  NaN                      NaN   
3               NaN                  NaN                      NaN   
4               NaN                  NaN                      NaN   

   retweet_count  retweeted  retweeted_status  \
0           8339      False               NaN   
1           6162      False               NaN   
2           4078      False               NaN   
3           8484      False               NaN   
4           9168      False               NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

<a id='assess'></a>
### Part II - Assessing  Data

After gathering each of the above pieces of data, our task is to assess them visually and programmatically for quality and tidiness issues. We should detect and document at least eight (8) quality issues and two (2) tidiness issues in the wrangle_act.ipynb Jupyter Notebook.

#### Step 1: Detect

**a. Visual Assesment**

In [8]:
# Visual assesment of the 1st dataframe: twitter-archive-enhanced.csv
twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [9]:
# Visual assesment of the 2nd dataframe: image-predictions.tsv
twitter_images

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [10]:
# Visual assesment of the 3rd dataframe: tweet_json.txt
twitter_json

contributors  coordinates                      created_at  \
0              NaN          NaN  Tue Aug 01 16:23:56 +0000 2017   
1              NaN          NaN  Tue Aug 01 00:17:27 +0000 2017   
2              NaN          NaN  Mon Jul 31 00:18:03 +0000 2017   
3              NaN          NaN  Sun Jul 30 15:58:51 +0000 2017   
4              NaN          NaN  Sat Jul 29 16:00:24 +0000 2017   
5              NaN          NaN  Sat Jul 29 00:08:17 +0000 2017   
6              NaN          NaN  Fri Jul 28 16:27:12 +0000 2017   
7              NaN          NaN  Fri Jul 28 00:22:40 +0000 2017   
8              NaN          NaN  Thu Jul 27 16:25:51 +0000 2017   
9              NaN          NaN  Wed Jul 26 15:59:51 +0000 2017   
10             NaN          NaN  Wed Jul 26 00:31:25 +0000 2017   
11             NaN          NaN  Tue Jul 25 16:11:53 +0000 2017   
12             NaN          NaN  Tue Jul 25 01:55:32 +0000 2017   
13             NaN          NaN  Tue Jul 25 00:10:02 +0000 2017   
14             NaN          NaN  Mon Jul 24 17:02:04 +0000 2017   
15             NaN          NaN  Mon Jul 24 00:19:32 +0000 2017   
16             NaN          NaN  Sun Jul 23 00:22:39 +0000 2017   
17             NaN          NaN  Sat Jul 22 16:56:37 +0000 2017   
18             NaN          NaN  Sat Jul 22 00:23:06 +0000 2017   
19             NaN          NaN  Thu Jul 20 16:49:33 +0000 2017   
20             NaN          NaN  Wed Jul 19 16:06:48 +0000 2017   
21             NaN          NaN  Wed Jul 19 03:39:09 +0000 2017   
22             NaN          NaN  Wed Jul 19 00:47:34 +0000 2017   
23             NaN          NaN  Tue Jul 18 16:08:03 +0000 2017   
24             NaN          NaN  Tue Jul 18 00:07:08 +0000 2017   
25             NaN          NaN  Mon Jul 17 16:17:36 +0000 2017   
26             NaN          NaN  Sun Jul 16 23:58:41 +0000 2017   
27             NaN          NaN  Sun Jul 16 20:14:00 +0000 2017   
28             NaN          NaN  Sat Jul 15 23:25:31 +0000 2017   
29             NaN          NaN  Sat Jul 15 16:51:35 +0000 2017   
...            ...          ...                             ...   
2326           NaN          NaN  Mon Nov 16 03:44:34 +0000 2015   
2327           NaN          NaN  Mon Nov 16 03:22:39 +0000 2015   
2328           NaN          NaN  Mon Nov 16 02:38:37 +0000 2015   
2329           NaN          NaN  Mon Nov 16 01:59:36 +0000 2015   
2330           NaN          NaN  Mon Nov 16 01:52:02 +0000 2015   
2331           NaN          NaN  Mon Nov 16 01:22:45 +0000 2015   
2332           NaN          NaN  Mon Nov 16 01:01:59 +0000 2015   
2333           NaN          NaN  Mon Nov 16 00:55:59 +0000 2015   
2334           NaN          NaN  Mon Nov 16 00:49:46 +0000 2015   
2335           NaN          NaN  Mon Nov 16 00:35:11 +0000 2015   
2336           NaN          NaN  Mon Nov 16 00:30:50 +0000 2015   
2337           NaN          NaN  Mon Nov 16 00:24:50 +0000 2015   
2338           NaN          NaN  Mon Nov 16 00:04:52 +0000 2015   
2339           NaN          NaN  Sun Nov 15 23:21:54 +0000 2015   
2340           NaN          NaN  Sun Nov 15 23:05:30 +0000 2015   
2341           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2342           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2343           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2344           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2345           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2346           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2347           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2348           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2349           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2350           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2351           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2352           NaN          NaN  Sun Nov 15 22:32:08 +0000 2015   
2353           NaN          NaN  Sun Nov 15 22:32:08 

**a. Programmatic Assessment**

In [11]:
# View 10 random records from the twitter_archive dataframe
twitter_archive.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
794   773336787167145985                    NaN                  NaN   
2258  667724302356258817                    NaN                  NaN   
2260  667550882905632768                    NaN                  NaN   
1536  689905486972461056                    NaN                  NaN   
371   828408677031882754                    NaN                  NaN   
85    876120275196170240                    NaN                  NaN   
65    879492040517615616                    NaN                  NaN   
804   772117678702071809                    NaN                  NaN   
701   786051337297522688           7.727430e+17         7.305050e+17   
248   845397057150107648                    NaN                  NaN   

                      timestamp  \
794   2016-09-07 01:47:12 +0000   
2258  2015-11-20 15:20:54 +0000   
2260  2015-11-20 03:51:47 +0000   
1536  2016-01-20 20:21:00 +0000   
371   2017-02-06 01:03:14 +0000   
85    2017-06-17 16:52:05 +0000   
65    2017-06-27 00:10:17 +0000   
804   2016-09-03 17:02:54 +0000   
701   2016-10-12 03:50:17 +0000   
248   2017-03-24 22:08:59 +0000   

                                                 source  \
794   <a href="http://twitter.com/download/iphone" r...   
2258  <a href="http://twitter.com" rel="nofollow">Tw...   
2260  <a href="http://twitter.com" rel="nofollow">Tw...   
1536  <a href="http://twitter.com/download/iphone" r...   
371   <a href="http://twitter.com/download/iphone" r...   
85    <a href="http://twitter.com/download/iphone" r...   
65    <a href="http://twitter.com/download/iphone" r...   
804   <a href="http://twitter.com/download/iphone" r...   
701   <a href="http://twitter.com/download/iphone" r...   
248   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
794   RT @dog_rates: Meet Fizz. She thinks love is a...         7.713808e+17   
2258  What a dog to start the day with. Very calm. L...                  NaN   
2260  RT @dogratingrating: Unoriginal idea. Blatant ...         6.675484e+17   
1536  Say hello to Gizmo. He's quite the pupper. Con...                  NaN   
371   This is Bear. He went outside to play in the s...                  NaN   
85    Meet Venti, a seemingly caffeinated puppoccino...                  NaN   
65    This is Bailey. He thinks you should measure e...                  NaN   
804   Meet Jack. He's a Clemson pup. Appears to be r...                  NaN   
701             13/10 for breakdancing puppo @shibbnbot                  NaN   
248   Say hello to Mimosa. She's an emotional suppor...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
794               4.196984e+09  2016-09-01 16:14:48 +0000   
2258                       NaN                        NaN   
2260              4.296832e+09  2015-11-20 03:41:59 +0000   
1536                       NaN                        NaN   
371                        NaN                        NaN   
85                         NaN                        NaN   
65                         NaN                        NaN   
804                        NaN                        NaN   
701                        NaN                        NaN   
248                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
794   https://twitter.com/dog_rates/status/771380798...                11   
2258  https://twitter.com/dog_rates/status/667724302...                 7   
2260  https://twitter.com/dogratingrating/status/667...                 5   
1536  https://twitter.com/dog_rates/status/689905486...                11   
371   https://twitter.com/dog_rates/status/828408677...                12   
85    https://twitter.com/dog_rates/status/876120275...                13   
65    https://twitter.com/dog_rates/status/879492040...                12   
804   https://twit

In [12]:
# View info of twitter_archive dataframe
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [13]:
# View descriptive statistics of twitter_archive dataframe
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [14]:
# View 10 random records from the twitter_images dataframe
twitter_images.sample(10)

tweet_id                                          jpg_url  \
1941  861005113778896900  https://pbs.twimg.com/media/C_LnlF5VoAEsL1K.jpg   
326   671891728106971137  https://pbs.twimg.com/media/CVMJ9guXAAAhAiK.jpg   
1722  819711362133872643  https://pbs.twimg.com/media/C2AzHjQWQAApuhf.jpg   
1109  723912936180330496  https://pbs.twimg.com/media/Cgva-QqUUAA7Hv9.jpg   
412   673919437611909120  https://pbs.twimg.com/media/CVo-JuMWwAAet6F.jpg   
1720  819347104292290561  https://pbs.twimg.com/media/C17n1nrWQAIErU3.jpg   
906   700167517596164096  https://pbs.twimg.com/media/Cbd-o8hWwAE4OFm.jpg   
1100  720775346191278080  https://pbs.twimg.com/media/CgC1WqMW4AI1_N0.jpg   
1803  832040443403784192  https://pbs.twimg.com/media/Cq9guJ5WgAADfpF.jpg   
572   678410210315247616  https://pbs.twimg.com/media/CWoyfMiWUAAmGdd.jpg   

      img_num                  p1   p1_conf  p1_dog                       p2  \
1941        1     German_shepherd  0.507951    True                 Pembroke   
326         1  Labrador_retriever  0.567933    True         golden_retriever   
1722        2        acorn_squash  0.848704   False              toilet_seat   
1109        1             Samoyed  0.991772    True               Pomeranian   
412         1     jack-o'-lantern  0.172079   False               schipperke   
1720        3          Rottweiler  0.909106    True  black-and-tan_coonhound   
906         1              beagle  0.162585    True                 Pembroke   
1100        1        Newfoundland  0.489970    True              groenendael   
1803        1  miniature_pinscher  0.796313    True                Chihuahua   
572         1          schipperke  0.145877    True       Labrador_retriever   

       p2_conf  p2_dog                         p3   p3_conf  p3_dog  
1941  0.136113    True                     muzzle  0.075764   False  
326   0.349401    True                  seat_belt  0.069396   False  
1722  0.044348   False                 toy_poodle  0.022009    True  
1109  0.003626    True                       chow  0.002232    True  
412   0.115984    True         miniature_pinscher  0.052175    True  
1720  0.044120    True                   Doberman  0.031835    True  
906   0.120481    True             Siberian_husky  0.110284    True  
1100  0.174497    True            giant_schnauzer  0.079067    True  
1803  0.155413    True  Staffordshire_bullterrier  0.030943    True  
572   0.098354    True                     kelpie  0.097393    True

In [15]:
# View info of twitter_images dataframe
twitter_images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [16]:
# View descriptive statistics of twitter_json dataframe
twitter_images.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [17]:
# View 10 random records from the twitter_json dataframe
twitter_json.sample(10)

contributors  coordinates                      created_at  \
151            NaN          NaN  Sun May 07 18:36:02 +0000 2017   
185            NaN          NaN  Sat Apr 22 18:31:02 +0000 2017   
1699           NaN          NaN  Fri Dec 25 19:39:43 +0000 2015   
373            NaN          NaN  Thu Feb 02 17:00:17 +0000 2017   
2084           NaN          NaN  Sat Nov 28 22:15:21 +0000 2015   
1196           NaN          NaN  Thu Mar 31 23:22:53 +0000 2016   
1465           NaN          NaN  Sat Jan 30 23:51:19 +0000 2016   
1845           NaN          NaN  Sat Dec 12 02:07:14 +0000 2015   
100            NaN          NaN  Wed Jun 07 16:14:40 +0000 2017   
2257           NaN          NaN  Fri Nov 20 00:12:54 +0000 2015   

     display_text_range                                           entities  \
151            [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
185            [0, 134]  {'hashtags': [{'text': 'ScienceMarch', 'indice...   
1699           [0, 115]  {'hashtags': [], 'symbols': [], 'user_mentions...   
373            [0, 129]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2084           [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1196            [0, 68]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1465           [0, 118]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1845           [0, 136]  {'hashtags': [], 'symbols': [], 'user_mentions...   
100            [0, 133]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2257           [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  favorite_count  \
151   {'media': [{'id': 861288473281437696, 'id_str'...           17307   
185   {'media': [{'id': 855851444666236933, 'id_str'...           46211   
1699  {'media': [{'id': 680473002811744256, 'id_str'...            2674   
373   {'media': [{'id': 827199955093843969, 'id_str'...           11203   
2084  {'media': [{'id': 670727697677606912, 'id_str'...             846   
1196  {'media': [{'id': 715680780459098112, 'id_str'...            4530   
1465                                                NaN            1727   
1845  {'media': [{'id': 675497072129249285, 'id_str'...            3241   
100   {'media': [{'id': 872486967967006720, 'id_str'...           40210   
2257  {'media': [{'id': 667495796489633792, 'id_str'...             525   

      favorited                                          full_text  geo  \
151       False  HI. MY. NAME. IS. BOOMER. AND. I. WANT. TO. SA...  NaN   
185       False  Here's a puppo participating in the #ScienceMa...  NaN   
1699      False  This is Hector. He thinks he's a hammer. Silly...  NaN   
373       False  This is Charlie. He wins every game of chess h...  NaN   
2084      False  This is Jeffrie. He's a handheld pup. Excellen...  NaN   
1196      False  This is Zeus. He's downright fabulous. 12/10 h...  NaN   
1465      False  Personally I'd give him an 11/10. Not sure why...  NaN   
1845      False  Meet Reggie. He's going for the world record. ...  NaN   
100       False  We. Only. Rate. Dogs. Do not send in other thi...  NaN   
2257      False  This is Philippe from Soviet Russia. Commandin...  NaN   

                            ...                          quoted_status  \
151                         ...                                    NaN   
185                         ...                                    NaN   
1699                        ...                                    NaN   
373                         ...                                    NaN   
2084                        ...                                    NaN   
1196                        ...                                    NaN   
1465                        ...                                    NaN   
1845                        ...                                    NaN   
100                         ...                                    NaN   
2257             

In [18]:
# View info of twitter_json dataframe
twitter_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 32 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2356 non-null object
display_text_range               2356 non-null object
entities                         2356 non-null object
extended_entities                2082 non-null object
favorite_count                   2356 non-null int64
favorited                        2356 non-null bool
full_text                        2356 non-null object
geo                              0 non-null float64
id                               2356 non-null int64
id_str                           2356 non-null int64
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null float64
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 non-null 

In [19]:
# View descriptive statistics of twitter_json dataframe
twitter_json.describe()

contributors  coordinates  favorite_count  geo            id  \
count           0.0          0.0     2356.000000  0.0  2.356000e+03   
mean            NaN          NaN     7949.753396  NaN  7.417684e+17   
std             NaN          NaN    12325.762302  NaN  6.837209e+16   
min             NaN          NaN        0.000000  NaN  6.660209e+17   
25%             NaN          NaN     1397.250000  NaN  6.776996e+17   
50%             NaN          NaN     3454.000000  NaN  7.178159e+17   
75%             NaN          NaN     9757.500000  NaN  7.986755e+17   
max             NaN          NaN   164611.000000  NaN  8.924206e+17   

             id_str  in_reply_to_status_id  in_reply_to_status_id_str  \
count  2.356000e+03           7.700000e+01               7.700000e+01   
mean   7.417684e+17           7.440692e+17               7.440692e+17   
std    6.837209e+16           7.524295e+16               7.524295e+16   
min    6.660209e+17           6.658147e+17               6.658147e+17   
25%    6.776996e+17           6.757073e+17               6.757073e+17   
50%    7.178159e+17           7.032559e+17               7.032559e+17   
75%    7.986755e+17           8.233264e+17               8.233264e+17   
max    8.924206e+17           8.862664e+17               8.862664e+17   

       in_reply_to_user_id  in_reply_to_user_id_str  quoted_status_id  \
count         7.700000e+01             7.700000e+01      2.600000e+01   
mean          2.040329e+16             2.040329e+16      8.113972e+17   
std           1.260797e+17             1.260797e+17      6.295843e+16   
min           1.185634e+07             1.185634e+07      6.721083e+17   
25%           3.589728e+08             3.589728e+08      7.761338e+17   
50%           4.196984e+09             4.196984e+09      8.281173e+17   
75%           4.196984e+09             4.196984e+09      8.637581e+17   
max           8.405479e+17             8.405479e+17      8.860534e+17   

       quoted_status_id_str  retweet_count  
count          2.600000e+01    2356.000000  
mean           8.113972e+17    2927.556876  
std            6.295843e+16    4936.906350  
min            6.721083e+17       0.000000  
25%            7.761338e+17     582.000000  
50%            8.281173e+17    1362.000000  
75%            8.637581e+17    3414.250000  
max            8.860534e+17   83874.000000

**Identify Duplicates for important data points**

In [20]:
# Investigate if duplicated Tweet IDs exist
twitter_archive.tweet_id.duplicated().sum()

0

In [21]:
twitter_images.tweet_id.duplicated().sum()

0

In [22]:
twitter_json.id.duplicated().sum()

14

In [23]:
# Identify duplicated IDs
twitter_json.id.value_counts()

666020888022790149    15
743510151680958465     1
825120256414846976     1
769212283578875904     1
700462010979500032     1
780858289093574656     1
699775878809702401     1
880095782870896641     1
760521673607086080     1
749075273010798592     1
776477788987613185     1
691820333922455552     1
715696743237730304     1
714606013974974464     1
760539183865880579     1
813157409116065792     1
676430933382295552     1
798644042770751489     1
833722901757046785     1
741099773336379392     1
818259473185828864     1
670704688707301377     1
667160273090932737     1
674394782723014656     1
672082170312290304     1
670093938074779648     1
759923798737051648     1
809920764300447744     1
805487436403003392     1
838085839343206401     1
                      ..
870308999962521604     1
720775346191278080     1
879492040517615616     1
775733305207554048     1
667911425562669056     1
834209720923721728     1
825026590719483904     1
758405701903519748     1
668986018524233728     1


In [24]:
# Investigate if duplicated jpg_url exist
twitter_images.jpg_url.duplicated().sum()

66

In [25]:
# Identify duplicated urls
twitter_images.jpg_url.value_counts()

https://pbs.twimg.com/media/Cwx99rpW8AMk_Ie.jpg                                            2
https://pbs.twimg.com/ext_tw_video_thumb/807106774843039744/pu/img/8XZg1xW35Xp2J6JW.jpg    2
https://pbs.twimg.com/ext_tw_video_thumb/815965888126062592/pu/img/JleSw4wRhgKDWQj5.jpg    2
https://pbs.twimg.com/media/CsGnz64WYAEIDHJ.jpg                                            2
https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg                                            2
https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg                                            2
https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg                                            2
https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg                                            2
https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg                                            2
https://pbs.twimg.com/media/CiibOMzUYAA9Mxz.jpg                                            2
https://pbs.twimg.com/media/Ct72q9jWcAAhlnw.jpg                       

In [26]:
twitter_images[twitter_images['jpg_url'] == 'https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg']

tweet_id                                          jpg_url  \
1785  829374341691346946  https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   
1903  851953902622658560  https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   

      img_num                         p1   p1_conf  p1_dog  \
1785        1  Staffordshire_bullterrier  0.757547    True   
1903        1  Staffordshire_bullterrier  0.757547    True   

                                  p2  p2_conf  p2_dog  \
1785  American_Staffordshire_terrier  0.14995    True   
1903  American_Staffordshire_terrier  0.14995    True   

                            p3   p3_conf  p3_dog  
1785  Chesapeake_Bay_retriever  0.047523    True  
1903  Chesapeake_Bay_retriever  0.047523    True

In [27]:
twitter_images[twitter_images['jpg_url'] == 'https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg']

tweet_id                                          jpg_url  \
224   670319130621435904  https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1345  759159934323924993  https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   

      img_num             p1   p1_conf  p1_dog      p2   p2_conf  p2_dog  \
224         1  Irish_terrier  0.254856    True  briard  0.227716    True   
1345        1  Irish_terrier  0.254856    True  briard  0.227716    True   

                               p3   p3_conf  p3_dog  
224   soft-coated_wheaten_terrier  0.223263    True  
1345  soft-coated_wheaten_terrier  0.223263    True

#### Step 2: Document

**a. Quality**

**Twitter Archive**
- Missing name of dogs
- Incorrect dog names (for instance: 'such', 'a', 'an')
- The rating numerator and denominator have invalid numbers (both columns having values with min 0 and max > 100)
- in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id should be intergs instead of float
- timestamp and retweeted_status_timestamp should be datetime instead of object (string)
- Dataset contains retweets

**Twitter Images**
- Missing records: 2075 rows instead of 2356
- 66 tweet_ids have the same duplicated jpg_urls

**Twitter API**
- There are a few data type issues but for this project we need only 3 columns: Tweet ID, retweet count and favorite count
- Tweet ID 666020888022790149 is duplicated 15 times    

**b. Tidiness**

**Twitter Archive**
- Different stage of dogs in columns instead of rows

**Twitter API**
- For this project we need only 3 columns: Tweet ID, retweet count and favorite count. Drop the rest of columns.

<a id='clean'></a>
### Part III - Cleaning Data

The very first thing to do before any cleaning occurs is to make a copy of each piece of data. All of the cleaning operations will be conducted on this copy so you can still view the original dirty and/or messy dataset later.

In [28]:
# Create copies of original DataFrames
twitter_archive_clean = twitter_archive.copy()
twitter_images_clean = twitter_images.copy()
twitter_json_clean = twitter_json.copy()

In [29]:
# We only want original ratings (no retweets) that have images so
# remove retweets from twitter_archive
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.retweeted_status_id.isna()]

In [30]:
# Drop columns related to retweets
twitter_archive_clean.drop(columns=['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp' ], inplace=True)

**About 1. in_reply_to_status_id and 2. in_reply_to_user_id**
1. If the represented Tweet is a reply, this field will contain the integer representation of the original Tweet’s ID
2. If the represented Tweet is a reply, this field will contain the integer representation of the original Tweet’s author ID. This will not necessarily always be the user directly mentioned in the Tweet. 

In [31]:
twitter_archive_clean.in_reply_to_status_id.value_counts()

6.671522e+17    2
8.562860e+17    1
8.131273e+17    1
6.754971e+17    1
6.827884e+17    1
8.265984e+17    1
6.780211e+17    1
6.689207e+17    1
6.658147e+17    1
6.737159e+17    1
7.590995e+17    1
8.862664e+17    1
7.384119e+17    1
7.727430e+17    1
7.468859e+17    1
8.634256e+17    1
6.693544e+17    1
6.914169e+17    1
6.920419e+17    1
6.753494e+17    1
7.291135e+17    1
8.406983e+17    1
6.747400e+17    1
7.501805e+17    1
6.744689e+17    1
7.638652e+17    1
6.747934e+17    1
8.503288e+17    1
6.747522e+17    1
8.816070e+17    1
               ..
8.380855e+17    1
8.211526e+17    1
8.558616e+17    1
8.558585e+17    1
7.032559e+17    1
6.678065e+17    1
8.018543e+17    1
7.667118e+17    1
6.855479e+17    1
6.717299e+17    1
6.715610e+17    1
6.758457e+17    1
6.924173e+17    1
7.476487e+17    1
8.381455e+17    1
6.903413e+17    1
8.476062e+17    1
8.352460e+17    1
6.813394e+17    1
8.795538e+17    1
6.860340e+17    1
8.571567e+17    1
6.765883e+17    1
7.044857e+17    1
8.707262e+

In [32]:
twitter_archive_clean.in_reply_to_user_id.value_counts()

4.196984e+09    47
2.195506e+07     2
7.305050e+17     1
2.916630e+07     1
3.105441e+09     1
2.918590e+08     1
2.792810e+08     1
2.319108e+09     1
1.806710e+08     1
3.058208e+07     1
2.625958e+07     1
1.943518e+08     1
3.589728e+08     1
8.405479e+17     1
2.894131e+09     1
2.143566e+07     1
2.281182e+09     1
1.648776e+07     1
4.717297e+09     1
2.878549e+07     1
1.582854e+09     1
4.670367e+08     1
4.738443e+07     1
1.361572e+07     1
1.584641e+07     1
2.068372e+07     1
1.637468e+07     1
1.185634e+07     1
1.198989e+09     1
1.132119e+08     1
7.759620e+07     1
Name: in_reply_to_user_id, dtype: int64

**Conclusions**
- 47 out of 77 tweets were replies to the original post of the tweet's author id: 4196983835, meaning tweets of @dog_rates
- Since I do not find these fields essential for the dataset, I am going to drop them

In [33]:
# Drop columns related to tweets that are reply to an original tweet (mainly coming from @dog_rates)
twitter_archive_clean.drop(columns=['in_reply_to_status_id', 'in_reply_to_user_id'], inplace=True)

In [34]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2175 non-null int64
timestamp             2175 non-null object
source                2175 non-null object
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: int64(3), object(9)
memory usage: 220.9+ KB


In [35]:
# Change datatype of timestamp from object to timestamp
twitter_archive_clean.timestamp = pd.to_datetime(twitter_archive_clean.timestamp)

In [36]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2175 non-null int64
timestamp             2175 non-null datetime64[ns]
source                2175 non-null object
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: datetime64[ns](1), int64(3), object(8)
memory usage: 220.9+ KB


In [37]:
# Print column names for melt
columns = twitter_archive_clean.columns.tolist()
print(columns[:-4])

['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls', 'rating_numerator', 'rating_denominator', 'name']


In [38]:
# Check dog stage values before melting
print(twitter_archive_clean.doggo.value_counts())
print(twitter_archive_clean.floofer.value_counts())
print(twitter_archive_clean.pupper.value_counts())
print(twitter_archive_clean.puppo.value_counts())

None     2088
doggo      87
Name: doggo, dtype: int64
None       2165
floofer      10
Name: floofer, dtype: int64
None      1941
pupper     234
Name: pupper, dtype: int64
None     2150
puppo      25
Name: puppo, dtype: int64


In [39]:
# Melt dog stages into one column
twitter_archive_clean = pd.melt(twitter_archive_clean, id_vars=['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls', 'rating_numerator', 'rating_denominator', 'name'],
                           var_name='stage', value_name='dog_stage')

In [40]:
# Drop duplicates after pivoting
twitter_archive_clean = twitter_archive_clean.sort_values('dog_stage').drop_duplicates('tweet_id', keep = 'last')

In [42]:
# Check if we did not miss any status
twitter_archive_clean.dog_stage.value_counts()

None       1831
pupper      234
doggo        75
puppo        25
floofer      10
Name: dog_stage, dtype: int64

In [43]:
# Drop stage column
twitter_archive_clean.drop(columns=['stage'], inplace=True)

In [44]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 2095 to 7298
Data columns (total 9 columns):
tweet_id              2175 non-null int64
timestamp             2175 non-null datetime64[ns]
source                2175 non-null object
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
dog_stage             2175 non-null object
dtypes: datetime64[ns](1), int64(3), object(5)
memory usage: 169.9+ KB
